apm-agent-plugins/apm-grpc/apm-grpc-plugin/src/main/java/co/elastic/apm/agent/grpc/GrpcHelper.java

```diff
@@ -382,16 +384,38 @@ public void onClientCallCreationExit(@Nullable ClientCall<?, ?> clientCall, @Nul
      * @param realClientCall        the client call instance that represents the actual client call
      */
     public void replaceClientCallRegistration(ClientCall<?, ?> placeholderClientCall, ClientCall<?, ?> realClientCall) {
-        Span spanOfPlaceholder = delayedClientCallSpans.remove(placeholderClientCall);
+        // we cannot remove yet, because the span could have been ended already through ClientCall#start(), in which case
+        // it will be recycled ahead of time due to reference decrement when removed from the map
+        Span spanOfPlaceholder = delayedClientCallSpans.get(placeholderClientCall);
         if (spanOfPlaceholder != null) {
```

Comment:
[minor] we could return early  to avoid one level of `if` nesting to make this slightly easier to read.